File: frontend/src/utils/dataset-modality.ts
import type { DatasetModality, DatasetSampleType } from "@/types/dataset.types";

const modalityOrder: DatasetSampleType[] = ["image", "video", "text"];

export function normalizeDatasetModality(value: string | string[]): DatasetModality {
  const source = Array.isArray(value)
    ? value
    : String(value || "")
      .replace(/image_text/g, "image+text")
      .replace(/[|/, ]+/g, "+")
      .split("+");
  const unique = new Set(
    source
      .map((item) => String(item).trim().toLowerCase())
      .filter((item): item is DatasetSampleType => ["image", "video", "text"].includes(item)),
  );
  return modalityOrder.filter((item) => unique.has(item)).join("+") as DatasetModality;
}

export function splitDatasetModality(value: string | null | undefined): DatasetSampleType[] {
  const normalized = normalizeDatasetModality(value || "");
  return normalized ? normalized.split("+") as DatasetSampleType[] : [];
}

export function datasetSupportsSampleType(modality: string | null | undefined, sampleType: DatasetSampleType): boolean {
  return splitDatasetModality(modality).includes(sampleType);
}

export function datasetModalityLabel(modality: string | null | undefined): string {
  const parts = splitDatasetModality(modality);
  if (!parts.length) return "-";
  const labels: Record<DatasetSampleType, string> = {
    image: "图片",
    video: "视频",
    text: "文本",
  };
  return parts.map((item) => labels[item]).join(" + ");
}

File: backend/app/core/dataset_modality.py
from __future__ import annotations

from typing import Iterable

from app.core.exceptions import ValidationError


DATASET_MODALITY_ORDER = ("image", "video", "text")
DATASET_MODALITY_SET = set(DATASET_MODALITY_ORDER)
LEGACY_DATASET_MODALITY_ALIASES = {
    "image_text": ("image", "text"),
}


def parse_dataset_modalities(value: str | Iterable[str] | None) -> tuple[str, ...]:
    if value is None:
        return ()
    if isinstance(value, str):
        raw = value.strip()
        if not raw:
            return ()
        if raw in LEGACY_DATASET_MODALITY_ALIASES:
            tokens = LEGACY_DATASET_MODALITY_ALIASES[raw]
        else:
            normalized = raw.replace(",", "+").replace("|", "+").replace("/", "+").replace(" ", "+")
            tokens = tuple(part.strip().lower() for part in normalized.split("+") if part.strip())
    else:
        tokens = tuple(str(part).strip().lower() for part in value if str(part).strip())
    unique_tokens = {token for token in tokens if token}
    invalid = sorted(unique_tokens - DATASET_MODALITY_SET)
    if invalid:
        raise ValidationError(f"unsupported dataset modality: {', '.join(invalid)}")
    return tuple(token for token in DATASET_MODALITY_ORDER if token in unique_tokens)


def normalize_dataset_modality(value: str | Iterable[str] | None) -> str:
    tokens = parse_dataset_modalities(value)
    if not tokens:
        raise ValidationError("dataset modality cannot be empty")
    return "+".join(tokens)


def dataset_supports_sample_type(modality: str | Iterable[str] | None, sample_type: str) -> bool:
    sample = str(sample_type or "").strip().lower()
    if sample not in DATASET_MODALITY_SET:
        return False
    return sample in parse_dataset_modalities(modality)
